src/hip/models.pyi

```diff
@@ -0,0 +1,81 @@
+import typing
+
+HeadersType = typing.Union[
+    typing.Mapping[str, str],
+    typing.Mapping[bytes, bytes],
+    typing.Iterable[typing.Tuple[str, str]],
+    typing.Iterable[typing.Tuple[bytes, bytes]],
+]
+
+class SyncRequestData:
+    """Represents a synchronous request data object. Basically a wrapper around whatever
+    a user passes in via 'data', 'files', 'json', etc parameters. We can take bytes /
+    strings, iterators of bytes or strings, and files. We can also sub-class / implement
+    a file-like interface for things like multipart-formdata so it's possible to
+    rewind (unlike the current urllib3 interface).
+
+    Maybe in the future we can expose details about when an object can be sent via
+    'socket.sendfile()', etc. This would have to be communicated somehow to the
+    low-level backend streams.
+
+    When we're handed a file-like object we should take down the starting point
+    in the file via .tell() so we can rewind and put the pointer back after
+    the seek() call in '.content_length'
+
+    Ref: https://github.com/python-trio/urllib3/issues/135
+    """
+
+    def read(self, nbytes: int) -> bytes:
+        """Don't know what to call this? Grab some data from the pool of data
+        to be sent across the wire.
+        """
+    @property
+    def content_length(self) -> typing.Optional[int]:
+        """We can get a proper content-length for bytes, strings, file-like objects
+        that are opened in binary mode (is that detectable somehow?). If we hand
+        back 'None' from this property it means that the request should use
+        'Transfer-Encoding: chunked'
+        """
+    def rewind(self) -> None:
+        """This function rewinds the request data so that it can be retransmitted
+        in the case of a timeout/socket error on an idempotent request, a redirect,
+        etc so that the new request can be sent. This works for file-like objects
+        and bytes / strings (where it's a no-op).
+        """
```

Comment:
Should also add a `content_type` property so that can be passed to the request as a default if none is set.